File: csharp/src/DataStructures/DisjointSet.cs
namespace DSA.DataStructures;

/// <summary>
/// Disjoint Set (Union-Find) implementation with path compression and union by rank.
/// Time: O(α(n)) amortized for all operations (nearly constant)
/// Space: O(n)
/// </summary>
public class DisjointSet
{
    private readonly int[] _parent;
    private readonly int[] _rank;
    private int _count;

    public int Count => _count;
    public int Size => _parent.Length;

    public DisjointSet(int size)
    {
        _parent = new int[size];
        _rank = new int[size];
        _count = size;

        for (int i = 0; i < size; i++)
            _parent[i] = i;
    }

    public int Find(int x)
    {
        if (_parent[x] != x)
            _parent[x] = Find(_parent[x]); // Path compression
        return _parent[x];
    }

    public bool Union(int x, int y)
    {
        var rootX = Find(x);
        var rootY = Find(y);

        if (rootX == rootY)
            return false;

        // Union by rank
        if (_rank[rootX] < _rank[rootY])
        {
            _parent[rootX] = rootY;
        }
        else if (_rank[rootX] > _rank[rootY])
        {
            _parent[rootY] = rootX;
        }
        else
        {
            _parent[rootY] = rootX;
            _rank[rootX]++;
        }

        _count--;
        return true;
    }

    public bool Connected(int x, int y) => Find(x) == Find(y);
}

File: php/src/DataStructures/DisjointSet.php
<?php

declare(strict_types=1);

namespace DSA\DataStructures;

/**
 * Disjoint Set (Union-Find) implementation with path compression and union by rank.
 * Time: O(α(n)) amortized for all operations (nearly constant)
 * Space: O(n)
 */
class DisjointSet
{
    private array $parent;
    private array $rank;
    private int $count;

    public function __construct(int $size)
    {
        $this->parent = range(0, $size - 1);
        $this->rank = array_fill(0, $size, 0);
        $this->count = $size;
    }

    public function find(int $x): int
    {
        if ($this->parent[$x] !== $x) {
            $this->parent[$x] = $this->find($this->parent[$x]); // Path compression
        }
        return $this->parent[$x];
    }

    public function union(int $x, int $y): bool
    {
        $rootX = $this->find($x);
        $rootY = $this->find($y);

        if ($rootX === $rootY) {
            return false;
        }

        // Union by rank
        if ($this->rank[$rootX] < $this->rank[$rootY]) {
            $this->parent[$rootX] = $rootY;
        } elseif ($this->rank[$rootX] > $this->rank[$rootY]) {
            $this->parent[$rootY] = $rootX;
        } else {
            $this->parent[$rootY] = $rootX;
            $this->rank[$rootX]++;
        }

        $this->count--;
        return true;
    }

    public function connected(int $x, int $y): bool
    {
        return $this->find($x) === $this->find($y);
    }

    public function count(): int
    {
        return $this->count;
    }

    public function size(): int
    {
        return count($this->parent);
    }
}
